torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -143,66 +144,140 @@ class ProcessGroupNCCL : public ProcessGroup {
     }
 
-    // FutureNCCL's value is NCCL collective's outputs and callbacks were
-    // invoked inline by addCallback(), so markCompleted is not needed.
+    // FutureNCCL's value is NCCL collective's outputs or callback() output with
+    // callbacks being invoked inline by addCallbackWithStream(), so
+    // markCompleted is not needed.
     void markCompleted(at::IValue /* unused */) override {
       C10_THROW_ERROR(Error, "FutureNCCL::markCompleted is not supported.");
     }
 
-    // Just returns NCCL collective's outputs after WorkNCCL's wait returns.
+    // Just returns FutureNCCL's value.
     at::IValue value() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
     const at::IValue& constValue() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
-    // Add a callback to FutureNCCL. It invokes the callback inline after
-    // WorkNCCL's wait(). workCallbacks return a Future (not FutureNCCL).
-    void addCallback(std::function<void(void)> callback) override {
-      work_->wait();
+    // Adds a callback to FutureNCCL. It invokes the callback inline after
+    // synchronizing FutureNCCL's own cudaEvents with the stream that guards
+    // this callback. FutureNCCL then will return a FutureNCCL if this callback
+    // is executed without any errors. This new FutureNCCL's cudaEvents will
+    // record the callback's stream.
+    void addCallbackWithStream(
+        std::function<void(void)> callback,
+        const c10::cuda::CUDAStream stream,
+        std::shared_ptr<std::vector<at::cuda::CUDAEvent>> thenFutCudaEvents) {
+      (*cudaEvents_)[0].block(stream);
+      c10::OptionalStreamGuard streamGuard{c10::Stream(stream)};
       callback();
+      (*thenFutCudaEvents)[0].record(stream);
+    }
+
+    // We use addCallbackWithStream instead of addCallback.
+    void addCallback(std::function<void(void)> /* unused */) override {
+      C10_THROW_ERROR(
+          Error,
+          "FutureNCCL uses addCallbackWithStream instead of addCallback.");
+    }
+
+    // Adds a callback to the futureNCCL, and returns another FutureNCCL to hold
+    // the return value of the callback and new cudaEvents that recorded the
+    // stream that guards this callback.
+    c10::intrusive_ptr<Future> then(
+        std::function<at::IValue(void)> callback,
+        at::TypePtr type) override {
+      // Set a default fut, this will be returned in case of an error in the
+      // callback. If the callback is executed without any errors, fut will be
+      // replaced by a FutureNCCL.
+      auto fut = c10::make_intrusive<Future>(type);
+
+      // Get a new stream from pool that will guard the callback.
+      const c10::cuda::CUDAStream stream =
+          at::cuda::getStreamFromPool(deviceIndex_);
+      // Create a new cudaEvents object that will record callback's stream and
+      // used by the new FutureNCCL.
+      auto thenFutCudaEvents =
```

Comment:
Any reason it needs to be a `std::vector` if it only holds one CUDA event?

Reply:
Referring the discussion in https://github.com/pytorch/pytorch/pull/42869#discussion_r470333493.